litestar/middleware/response_cache.py

```diff
@@ -50,7 +51,11 @@ async def wrapped_send(message: Message) -> None:
                     messages.append(message)
 
                 if messages and message["type"] == HTTP_RESPONSE_BODY and not message["more_body"]:
-                    key = (route_handler.cache_key_builder or self.config.key_builder)(Request(scope))
+                    key_builder = route_handler.cache_key_builder or self.config.key_builder
+                    if asyncio.iscoroutinefunction(key_builder):
+                        key = await key_builder(Request(scope))
```

Comment:
In the linked issue (#2977) you access `request.json()` inside the key builder. That will fail here, because the request object that you pass the key builder isn't given the `receive` coroutine, so calling `await request.json()` won't work. You can test this out with `python -m asyncio`:

```python-console
asyncio REPL 3.8.18 (default, Oct 26 2023, 08:29:21) 
[GCC 11.4.0] on linux
Use "await" directly instead of "asyncio.run()".
Type "help", "copyright", "credits" or "license" for more information.
>>> import asyncio
>>> from litestar.connection import Request
>>> fake_scope = {"state":{}}
>>> await Request(fake_scope).json()
Traceback (most recent call last):
  File "/home/peter/.pyenv/versions/3.8.18/lib/python3.8/asyncio/__main__.py", line 58, in runcode
    return future.result()
  File "/home/peter/.pyenv/versions/3.8.18/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/home/peter/.pyenv/versions/3.8.18/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "<console>", line 1, in <module>
  File "/home/peter/PycharmProjects/litestar/litestar/connection/request.py", line 133, in json
    body = await self.body()
  File "/home/peter/PycharmProjects/litestar/litestar/connection/request.py", line 196, in body
    self._body = self._connection_state.body = b"".join([c async for c in self.stream()])
  File "/home/peter/PycharmProjects/litestar/litestar/connection/request.py", line 196, in <listcomp>
    self._body = self._connection_state.body = b"".join([c async for c in self.stream()])
  File "/home/peter/PycharmProjects/litestar/litestar/connection/request.py", line 167, in stream
    while event := await self.receive():
  File "/home/peter/PycharmProjects/litestar/litestar/connection/base.py", line 38, in empty_receive
    raise RuntimeError()
RuntimeError
>>> 
```
Awaiting `Request.json()` or any other method that would result in the `receive()` coroutine being awaited within middleware is generally not a good idea because other middleware in the stack might need to wrap `receive`, but it will be too late if it has already been awaited.